sw/device/tests/hmac/dif_hmac_test.c

```diff
@@ -0,0 +1,190 @@
+      case kDifHmacNoError:
+        break;
+      case kDifHmacInvalidArgument:
+        uart_send_str("Invalid arguments encountered checking FIFO depth.\r\n");
+        return false;
+    }
+  } while (fifo_depth > 0);
+
+  switch (dif_hmac_run(&hmac)) {
+    case kDifHmacNoError:
+      break;
+    case kDifHmacInvalidArgument:
+      uart_send_str("Invalid arguments encountered running HMAC.\r\n");
+      return false;
+  }
+
+  bool hmac_running;
+  do {
+    switch (dif_hmac_done(&hmac, &hmac_running)) {
+      case kDifHmacNoError:
+        break;
+      case kDifHmacInvalidArgument:
+        uart_send_str(
+            "Invalid arguments encountered checking HMAC run status.\r\n");
+        return false;
+    }
+  } while (hmac_running);
+
+  uint64_t sent_bits;
+  switch (dif_hmac_get_msg_len(&hmac, &sent_bits)) {
+    case kDifHmacNoError:
+      break;
+    case kDifHmacInvalidArgument:
+      uart_send_str(
+          "Invalid arguments encountered checking message length.\r\n");
+      return false;
+  }
+
+  if (sent_bits != len * 8) {
+    uart_send_str("Message length mismatch. Expected ");
+    print_uint(len * 8);
+    uart_send_str(" bits but got ");
+    print_uint(sent_bits);
+    uart_send_str(" bits.\r\n");
+  }
+
+  dif_hmac_digest_t digest_result;
+  switch (dif_hmac_digest_read(&hmac, &digest_result)) {
+    case kDifHmacNoError:
+      break;
+    case kDifHmacInvalidArgument:
+      uart_send_str("Invalid arguments encountered reading HMAC digest.\r\n");
+      return false;
+  }
+
+  bool matches = true;
+  for (int i = 0; i < 8; i++) {
+    if (expected_digest[i] != digest_result.digest[i]) {
+      uart_send_str("Digest mismatch, expected: 0x");
+      uart_send_uint(expected_digest[i], sizeof(uint32_t) * 8);
+      uart_send_str(" at index [");
+      print_uint(i);
+      uart_send_str("] but got: 0x");
+      uart_send_uint(digest_result.digest[i], sizeof(uint32_t) * 8);
+      uart_send_str(".\r\n");
+      matches = false;
+    }
+  }
+
+  return matches;
+}
+
+int main(int argc, char **argv) {
+  uart_init(UART_BAUD_RATE);
+  uart_send_str("Running HMAC DIF test\r\n");
+
+  // Test SHA256
+  uart_send_str("Testing SHA256 mode...\r\n");
+
+  dif_hmac_config_t hmac_config;
+  hmac_config.base_addr = mmio_region_from_addr(HMAC0_BASE_ADDR);
+  hmac_config.digest_endianness = kDifHmacBigEndian;
+  hmac_config.message_endianness = kDifHmacBigEndian;
+  hmac_config.mode = kDifHmacModeSha256;
+
+  if (run_test(&hmac_config, kData, sizeof(kData), kExpectedShaDigest)) {
+    uart_send_str("PASS!\r\n");
+  } else {
+    uart_send_str("FAIL!\r\n");
+  }
+
+  // Test HMAC
+  uart_send_str("\r\nTesting HMAC mode...\r\n");
+
+  hmac_config.mode = kDifHmacModeHmac;
+  for (int i = 0; i < 8; i++) {
+    hmac_config.key[i] = kHmacKey[i];
+  }
+  if (run_test(&hmac_config, kData, sizeof(kData), kExpectedHmacDigest)) {
+    uart_send_str("PASS!\r\n");
+  } else {
+    uart_send_str("FAIL!\r\n");
+  }
```

Comment:
A minor point but it would be cleaner if we had a wrapper around `run_test` that takes also a test name and handles success and failure internally. You can consider renaming `run_test` to `run_test_impl` and write `run_test()` like
```
void run_test(const char *test_name, const dif_hmac_config_t *config, const char *data,
             size_t len, const uint32_t *expected_digest) {
  uart_send_str("\r\n");
  uart_send_str(test_name);
  uart_send_str("\r\n");
  if (run_test(config, data, len, expected_digest) {
    uart_send_str("PASS!\r\n");
  } else {
    uart_send_str("FAIL!\r\n");
  }
}